File: 1676-lowest-common-ancestor-of-a-binary-tree-iv/1676-lowest-common-ancestor-of-a-binary-tree-iv.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode LCA = null;
    
    public int getLCA(TreeNode node,HashSet<Integer> set) {
        //Base Case
        if(node == null) return 0;
        //left and right recursive call
        int leftFound = getLCA(node.left,set);
        int rightFound = getLCA(node.right,set);
        //calculate totalFOund
        int totalFound = leftFound + rightFound;
        //check if current node is to be searched for
        if(set.contains(node.val)) totalFound++;
        
        if(LCA == null && totalFound == set.size()) LCA = node;
        //return totalfound to parent
        return totalFound;
    }
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode[] nodes) {
        HashSet<Integer> set = new HashSet<>();
        for(TreeNode node : nodes) set.add(node.val);
        getLCA(root,set);
        return LCA;
    }
}

File: 1676-lowest-common-ancestor-of-a-binary-tree-iv/1676-lowest-common-ancestor-of-a-binary-tree-iv.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* LCA = NULL;
    int getLCA(TreeNode* node,unordered_set<int>& set) {
        //Base Case
        if(!node) return 0;
        //left and right recursive call
        int leftFound = getLCA(node->left,set);
        int rightFound = getLCA(node->right,set);
        //calculate totalFOund
        int totalFound = leftFound+rightFound;
        //check if current node is to be searched for
        if(set.find(node->val) != set.end()) totalFound++;
        //check of current node is LCA
        if(LCA == NULL && totalFound == set.size()) LCA = node;
        //return totalfound to parent
        return totalFound;
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, vector<TreeNode*> &nodes) {
        unordered_set<int> set;
        for(TreeNode* node : nodes) set.insert(node->val);
        getLCA(root,set);
        return LCA;
    }
};